### ISSUE TITLE
[Linux] output=ttf issues

### ISSUE THREAD

**Issue by rderooy** [joncampbell123/dosbox-x#1999]
**Describe the bug**
The ``output=ttf`` option is a welcome addition, but there are a few issues that I encountered.

- No default ttf font set. I tried it with ``ttf.font=/usr/share/fonts/liberation-mono/LiberationMono-Regular.ttf``
- output ttf menu option greyed out, even if you set a font in the config file
- starting with ``-set output=ttf`` (after setting font), causes the console to take up the entire screen size (and then some due to title bar, menu, borders, etc) with no resize option from the menu. I had to set ``ttf.winperc`` to ~40 or less for it not to take up the entire screen. Perhaps a simple font size option would be more user friendly? The below screenshot is with ttf.winperc=1
- horizontal lines are not continuous, as already reported elsewhere. But also the vertical.
- menu bar is invisible until you hover over it with the mouse, and even then only the parts you have gone over, and the right hand side of it never becomes visible as can be seen in the screenshot. Basically it seems the dosbox-x output is too far up and overlaying the menu bar until you go over it with the mouse.
- changing from output=ttf to output=opengl using the menu causes a crash
```
LOG: E_Exit: SDL:Can't set palette
E_Exit: SDL:Can't set palette
```
![Screenshot from 2020-11-15 15-31-25](https://user-images.githubusercontent.com/17701098/99187572-a7747f80-2757-11eb-83a7-a01213919cc6.png)

**Comment by joncampbell123** [comment]
Also an issue that I am addressing is that SDL drawn menus do not yet work with this output (which are needed here on Linux).

**Comment by Wengier** [comment]
@rderooy There is a default TTF font, which is the built-in TTF font (activated when you leave the ttf.font option empty). But I will improve it so that when an external TTF font cannot be loaded it will fallback to the built-in TTF font.

As for the continuity of lines, I think it is really font-specific at this time. Try the default font or a different TTF font and see how it works.

The issue when changing from output=ttf to output=opengl using the menu seems to only occur with Linux SDL1 build. It appears to work fine with Windows SDL1/SDL2 builds as well as Linux SDL2 build.

The SDL drawn menu bar menu bar is indeed invisible until you hover over it with the mouse as you said.

**Comment by rderooy** [comment]
@Wengier your right, I missed the fact that there was a built-in font. With that lines look at lot better.

Also with the default font, the default ``ttf.winperc`` value no longer causes a window size that is basically too large for the screen. Still, the ``ttf.winperc`` setting has relatively little effect, the screenshot below is with ``sdl.winperc=1`` and is still quite large. Larger than a standard dosbox-x window with output set to surface or opengl.

- The menu bar situation is slightly improved with the latest code, there is now the space for a menu bar but it is still invisible until you mouse over it.
- starting the SDL2 version with output=ttf causes some momentary messed up ascii characters to display on startup.
- I can confirm that the crash when switching output only happens with the SDL1 version.

![Screenshot from 2020-11-16 08-07-36](https://user-images.githubusercontent.com/17701098/99222419-d257e580-27e2-11eb-8323-ea30de5914fc.png)

**Comment by Wengier** [comment]
I think I have essentially fixed all the remaining issues in pull request #1999.

1. Output TTF menu option should no longer be greyed out even if it is not the first output selection.

2. Config option ttf.ptsize is added which allows you to set a font point size (min size: 10) instead of windows size. With default 0 ttf.winperc will be used instead.

3. SDL drawn menu bar should be visible by default (after @joncampbell123's fix to make room for the menu bar).

4, Changing from output=ttf to output=opengl using the menu should no longer cause a crash in Linux SDL1 build.

There are more fixes in this pull request of course, but they are not covered in this issue.

**Comment by rderooy** [comment]
Thanks guys, great work. Most of issues are resolved with this.

The only remaining issue is the momentary corruption when starting the SDL2 version.

And one other issue that I just noticed when I do a ``mode co40`` the display gets messed up. Basically everything is output twice.

**Comment by joncampbell123** [comment]
It should handle 40-column just fine, I just worked on that last night, at least with SDL1 builds.

Furthermore modifications to @Wengier VGA draw code should support any "stride" in text mode in case DOS programs manually set video modes by directly programming CRTC registers.

**Comment by rderooy** [comment]
I just did another build, and this is what happens if I use ``mode co40``. Basically everything gets shown twice, once on the left and once on the right side of the screen. This happens with both SDL1 and SDL2.
![Screenshot from 2020-11-17 08-19-51](https://user-images.githubusercontent.com/17701098/99358456-da308c00-28ad-11eb-8f5d-f74bf1b4fcef.png)

In addition, as I mentioned, using the integrated screenshot function gives a black screen in ttf mode.
![command_000](https://user-images.githubusercontent.com/17701098/99358625-1532bf80-28ae-11eb-95d5-a6d323d69cc9.png)

**Comment by joncampbell123** [comment]
mode co40 and mode co80 is working fine here, with machine=svga_s3.

However I am getting that same error with machine=cga.

**Comment by rderooy** [comment]
The ``mode co40`` issue is with a newly generated config file (``config -wcd -all`` and restart of dosbox-x), so with ``machine=svga_s3``. ``mode co80`` switches back to 80 column without apparent issue.

But strangely enough, this does result in a 40 column display, although not with double width characters:
``dosbox-x -defaultconf -set output=ttf -c "mode co40"``
![Screenshot from 2020-11-17 09-05-11](https://user-images.githubusercontent.com/17701098/99362646-0cdd8300-28b4-11eb-98a6-f16de876bf02.png)

**Comment by rderooy** [comment]
If I start with either:
``dosbox-x -defaultconf -set output=ttf -c "mode co40"`` or ``dosbox-x -defaultconf -set output=ttf -c "mode co80"``, I can afterwards switch between 40 and 80 columns without issue.

But if I start with ``dosbox-x -defaultconf -set output=ttf`` and then type ``mode co40`` I get the issue where everything is printed twice.

**Comment by joncampbell123** [comment]
@rderooy If I use a dosbox.conf that sets output=ttf and machine=svga_s3 I don't get this problem, but I can reproduce the issue if output=ttf and machine=cga.

**Comment by Wengier** [comment]
@rderooy Can you try the latest code in my pull request? Does the said command work?

**Comment by Wengier** [comment]
Also, with the TTF output you can now toggle between windowed and full-screen mode, and it supports all machine types, along with other changes. Please check it out.

**Comment by rderooy** [comment]
@Wengier I just tried your PR with SDL1 and SDL2 builds, and it solves the issue for the default svga_s3 mode. I could switch to 40 columns without the double display issue.

The issue still remains however when using ``machine=cga``. e.g.:
``dosbox-x -defaultconf -set output=ttf -set machine=cga -c "mode co40"``

**Comment by Wengier** [comment]
@rderooy Right now the video changes are only for VGA machine types, not the CGA ones. I temporarily disabled a recent change to make it work, but the change is only made for VGA machine types. The CGA ones are in a separate condition.

**Comment by Wengier** [comment]
@rderooy Please try the latest code in my pull request. "MODE CO40" should now work in CGA mode too.

**Comment by rderooy** [comment]
@Wengier seems @joncampbell123 already pulled it ;-)

In any case, I just tried it and it solved the CGA issue.

**Comment by Wengier** [comment]
@rderooy Just want to let you know there is indeed resizing options from the menu - check out “Video” -> “Text mode”. You will find other option related to text-mode there too.

**Comment by rderooy** [comment]
@Wengier nice!

The only issues I still have are the following:

- By default DOSBox-X uses a double line border on the opening screen, but with TTF this is shown as a single line border.

- If you start with ``dosbox-x -set output=ttf`` and use the integrated screenshot option, the resulting screenshot will be completely black. But if you start in regular output=overlay and then switch using the menu to TTF, the screenshot option will work, but will not use TTF.

**Comment by Wengier** [comment]
I think the border appearance is specific to the built-in font. With the Lucon font I get the following:

![image](https://user-images.githubusercontent.com/8216923/99921230-0194b800-2cf7-11eb-8326-2febe5978380.png)

And with the Nouveau_IBM font:

![image](https://user-images.githubusercontent.com/8216923/99921192-c7c3b180-2cf6-11eb-8659-ef32f03ccbe1.png)

Apparently different fonts have different appearances.

Meanwhile, you are right about the built-in screenshot option.

**Comment by Wengier** [comment]
@rderooy Although it is not yet the full solution, but the integrated screenshot (and video capture) option will now indeed work when TTF output is active, by a temporary video switch. No longer completely black as you mentioned.

**Comment by dreamer** [comment]
Not sure if it was reported already, but TTF output in 0.83.8 is broken when going fullscreen:
![Screenshot from 2020-12-02 00-55-59](https://user-images.githubusercontent.com/3967/100810692-4c21dd00-3439-11eb-8226-871f160231fe.png)

**Comment by Wengier** [comment]
@dreamer Thanks for testing and reporting the issue you encountered. Since you use Linux, I assume you are using the official DOSBox-X 0.83.8 Linux Flatpak version. I have tried to run the official Linux Flatpak on my Linux VM with the config settings ```output=ttf``` and ```ttf.winperc=100``` (or with the option ```-fullscreen```/```-fs``` from the command-line), and I got the following:

![image](https://user-images.githubusercontent.com/8216923/100812347-cdae4680-340a-11eb-8e15-9fad0d38dbdf.png)

The result seems to be consistent as I tried many times (at least 20). But it is apparently not a 100% full-screen since there is a small part of the screen not covered by the DOSBox-X window.

Switching from the windowed-mode to full-screen mode via the menu (or using a shortcut) also seems to work the same as above.

However, if I start the DOSBox-X Flatpak with the config settings ```output=ttf``` and ```fullscreen=true```, I may occasionally (not always) get a semi-frozen screen, which I have not encountered with the above-mentioned options, even though ```fullscreen=true``` is supposed to be the same as the command line option ```-fullscreen```/```-fs```. So there is something strange here which I will look further. In any case, thanks for letting me aware of such issue (although the screen you got seems to look somewhat different, which may be specific to the settings or so).

**Comment by dreamer** [comment]
Yup, I was testing Flatpak version - testing with X11, stock Gnome 3.36.7 (Fedora 32), dual-screen setup, Mesa 20.2.2. - the exact test steps were:

1. Start DOSBox-X via Gnome GUI (it starts on my default screen - right one)
2. Click Video -> Output -> TrueType font
3. Press F12+F to go fullscreen

**Comment by Wengier** [comment]
Thanks for checking it out! Since the issue does not seem to occur in Windows and macOS, I think it may be specific to Linux, and it actually also depends on the display server being used as you have mentioned - whether it is Wayland or not. I will test the different display server in my Linux VM and see what will happen. Thanks for letting me aware of this!

Meanwhile, I have made a PR to fix the TTF full-screen not fully cover the screen issue which I found earlier.
